### DISCUSSION TITLE
Is there a better solution for parametrised code blocks in markdown?

### DISCUSSION THREAD

**Discussion by ilg-ul** [facebook/docusaurus#10312] [Q&A]
Since most of the code blocks I have are parametrised (the code is shared between multiple sites),  I ended up with constructs like this:

```
<CodeBlock language="sh"> {
`xpm run install -C ~/Work/xpack-dev-tools/${ customField('appLcName') }-xpack.git && \\
xpm run link-deps -C ~/Work/xpack-dev-tools/${ customField('appLcName') }-xpack.git && \\
xpm run docker-prepare --config linux-arm -C ~/Work/xpack-dev-tools/${ customField('appLcName') }-xpack.git && \\
xpm run docker-link-deps --config linux-arm -C ~/Work/xpack-dev-tools/${ customField('appLcName') }-xpack.git`
}</CodeBlock>
```

In other words, instead of markdown I use the `<CodeBlock>` component, with a single JavaScript template string.

It is functional, but it is more tedious to write than usual markdown.

Is there a better solution?

**Comment by slorber** [comment]
How would you do this in pure JS otherwise? (I mean not involving Docusaurus)

You can implement a remark plugin that process strings wrapped in `<CodeBlock>` and replace `%%fieldName%%` placeholder by its value.

That's not std Markdown and you'd have to implement this yourself. However we might implement this as a first-class feature in the future. 

Related to https://github.com/facebook/docusaurus/issues/395, also providing a possible implementation.

**Reply by ilg-ul** [reply]
I don't have a solution, but I think that most people would expect a syntax like `{ expression }` to be accepted in code blocks, as it is in regular paragraphs.

**Reply by slorber** [reply]
We don't define the syntax, CommonMark + MDX does. 

You can extend the syntax and add non-standard behavior thanks to a remark/rehype plugins and micromark extension (for example: [micromark-extension-directive](https://github.com/micromark/micromark-extension-directive) adds support for parsing things like `:::note`). 

> most people would expect a syntax like `{ expression }` to be accepted in code blocks

Actually, this syntax is supported:

```js
<CodeBlock>
  {`my code`}
</CodeBlock>
```

The expression content is standard JS. 

Unfortunately for you, in std JS, using `{}` inside a string does not do anything, and you have to use template literal strings with "`"

```js
const age = 42;

// This prints "Age = {placeholder}"
"Age = {placeholder}"

// This prints "Age = 42"
`Age = ${placeholder}`
```

Note that template literal strings are probably more powerful than you think, in particular tagged template literals: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates

Instead of writing 

```js
<CodeBlock>
  {`my age is ${customFields.age}`}
</CodeBlock>
```

You could write:

```js
<CodeBlock>
  {myTag`my age is ${'age'}`}
</CodeBlock>
```

---

If we ever implement support for variable/placeholder substitution in Markdown files (https://github.com/facebook/docusaurus/issues/395), it is unlikely that we will use `{}` syntax. We don't want to create ambiguity, a substituted placeholder is not similar to a std JS expression, so it's probably better to explicitly invent a different syntax.

````mdx
<CodeBlock>
  {`my age is %age%`}
</CodeBlock>

```js
const age = %age%;
```
````

Note: you can already swizzle the CodeBlock component today and enhance it with your custom variable substitution logic: `props.children.replace("%age%",42)`

**Comment by ilg-ul** [comment]
> I think that most people would expect a syntax like { expression } to be accepted in code blocks

By 'code block' I understand markdown code blocks.

``````
```sh
xpm install @xpack-dev-tools/{ customField('appLcName') }@latest --verbose
```
``````

The `{ expression }` syntax is fine, and is accepted in most markdown constructs, with some exceptions (code, code block and links being the notable ones).

Like it or not, these exceptions make converting sites from Jekyll to Docusaurus a pretty tedious job. And the resulting text (or rather should I call it _code_) is a lot more difficult to read.

**Reply by slorber** [reply]
> By 'code block' I understand markdown code blocks.

`{}` is a valid char inside a string and a markdown code block is an unprocessed string. Imagine someone is using a language for which `{}` have a meaning (unlike sh). For example... JS!

````md
```jsx
const age = 42;
console.log(`my age is ${age}`)
<div>{age}</div>
```
````

Do you expect it to print `my age is 42`, or do you expect it to print `my age is <customField.age>`? 

There's an ambiguity here. We don't want to "override" and run some magic on top of what JS is already doing, altering the standard behavior of the underlying language while the user is trying to use `{}` in a standard way.

Maybe this makes migration from Jekyll harder, but not implementing such magic makes it also way easier to understand for any dev familiar with React/JS/MDX, which is primarily our target audience for Docusaurus.

> And the resulting text (or rather should I call it code) is a lot more difficult to read.

I'm pretty sure you'd find a way to achieve your need if you were more willing to learn the fundamentals of JS, React, and MDX. As I have shown you, there are many ways you can simplify it, and you'll only be able to find your path if you know enough of the underlying technologies and the pros/cons compared to Ruby/Jekyll.

We have already discussed before but Jekyll has an easier job to do than Docusaurus considering it's only rendering things on the server.
